drivers/usb_c/tcpc/ucpd_stm32.c

```diff
@@ -1348,10 +1348,12 @@ static const struct tcpc_driver_api driver_api = {
 #define DEV_INST_INIT(n) dev_inst[n] = DEVICE_DT_INST_GET(n);
 static void config_tcpc_irq(void)
 {
+	static int inst_num;
 	static const struct device
 	*dev_inst[DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)];
 
-	if (init_irq) {
+	/* Initialize and enable shared irq on last instance */
+	if (init_irq && ++inst_num == DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)) {
```

Comment:
You don't need both `init_irq` and `inst_num`.

Also, the caller to `config_tcpc_irq()` already has the `struct device *` pointer.  So the caller can check

```
if (dev == DEVICE_DT_INST_GET(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)) {
    config_tcpc_irq();
}
```

Reply:
I removed the init_irq flag but the other code change results in a build error, probably because the way DEVICE_DT_INST_GET is processed:
```
if (dev == DEVICE_DT_INST_GET(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT)) {
    config_tcpc_irq();
}
```
 /home/sam/zephyrproject/zephyr/drivers/usb_c/tcpc/ucpd_stm32.c
In file included from /home/sam/zephyrproject/zephyr/include/zephyr/toolchain/gcc.h:89,
                 from /home/sam/zephyrproject/zephyr/include/zephyr/toolchain.h:50,
                 from /home/sam/zephyrproject/zephyr/lib/libc/minimal/include/string.h:13,
                 from /home/sam/zephyrproject/zephyr/include/zephyr/sys/mpsc_packet.h:9,
                 from /home/sam/zephyrproject/zephyr/include/zephyr/logging/log_msg.h:10,
                 from /home/sam/zephyrproject/zephyr/include/zephyr/logging/log_core.h:9,
                 from /home/sam/zephyrproject/zephyr/include/zephyr/logging/log.h:11,
                 from /home/sam/zephyrproject/zephyr/drivers/usb_c/tcpc/ucpd_stm32.c:9:
/home/sam/zephyrproject/zephyr/drivers/usb_c/tcpc/ucpd_stm32.c: In function 'ucpd_isr_init':
/home/sam/zephyrproject/zephyr/include/zephyr/device.h:83:41: error: '__device_dts_ord_DT_N_INST_2_st_stm32_ucpd_ORD' undeclared (first use in this function)
   83 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
/home/sam/zephyrproject/zephyr/include/zephyr/device.h:209:37: note: in expansion of macro 'DEVICE_NAME_GET'
  209 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
/home/sam/zephyrproject/zephyr/include/zephyr/device.h:226:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  226 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/home/sam/zephyrproject/zephyr/include/zephyr/device.h:237:34: note: in expansion of macro 'DEVICE_DT_GET'
  237 | #define DEVICE_DT_INST_GET(inst) DEVICE_DT_GET(DT_DRV_INST(inst))
      |                                  ^~~~~~~~~~~~~
/home/sam/zephyrproject/zephyr/drivers/usb_c/tcpc/ucpd_stm32.c:1352:20: note: in expansion of macro 'DEVICE_DT_INST_GET'
 1352 |         if (dev == DEVICE_DT_INST_GET(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT))) {
      |                    ^~~~~~~~~~~~~~~~~~
/home/sam/zephyrproject/zephyr/include/zephyr/device.h:83:41: note: each undeclared identifier is reported only once for each function it appears in
   83 | #define DEVICE_NAME_GET(dev_id) _CONCAT(__device_, dev_id)
      |                                         ^~~~~~~~~
/home/sam/zephyrproject/zephyr/include/zephyr/device.h:209:37: note: in expansion of macro 'DEVICE_NAME_GET'
  209 | #define DEVICE_DT_NAME_GET(node_id) DEVICE_NAME_GET(Z_DEVICE_DT_DEV_ID(node_id))
      |                                     ^~~~~~~~~~~~~~~
/home/sam/zephyrproject/zephyr/include/zephyr/device.h:226:34: note: in expansion of macro 'DEVICE_DT_NAME_GET'
  226 | #define DEVICE_DT_GET(node_id) (&DEVICE_DT_NAME_GET(node_id))
      |                                  ^~~~~~~~~~~~~~~~~~
/home/sam/zephyrproject/zephyr/include/zephyr/device.h:237:34: note: in expansion of macro 'DEVICE_DT_GET'
  237 | #define DEVICE_DT_INST_GET(inst) DEVICE_DT_GET(DT_DRV_INST(inst))
      |                                  ^~~~~~~~~~~~~
/home/sam/zephyrproject/zephyr/drivers/usb_c/tcpc/ucpd_stm32.c:1352:20: note: in expansion of macro 'DEVICE_DT_INST_GET'
 1352 |         if (dev == DEVICE_DT_INST_GET(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT))) {
      |                    ^~~~~~~~~~~~~~~~~~
[168/181] Building C object zephyr/kernel/CMakeFiles/kernel.dir/xip.c.obj
ninja: build stopped: subcommand failed.
```